driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,364 @@
+    }
+
+    class PerHostErrorTracker implements LatencyTracker {
+
+        private final int maxErrorsPerMinute;
+        private final ErrorFilter errorFilter;
+        private final Clock clock;
+        private final ConcurrentMap<Host, RollingCount> hostsCounts = new ConcurrentHashMap<Host, RollingCount>();
+        private final ConcurrentMap<Host, Long> exclusionTimes = new ConcurrentHashMap<Host, Long>();
+
+        PerHostErrorTracker(int maxErrorsPerMinute, ErrorFilter errorFilter, Clock clock) {
+            this.maxErrorsPerMinute = maxErrorsPerMinute;
+            this.errorFilter = errorFilter;
+            this.clock = clock;
+        }
+
+        @Override
+        public void update(Host host, Statement statement, Exception exception, long newLatencyNanos) {
+            if (exception == null) {
+                return;
+            }
+            if (!errorFilter.shouldConsiderError(exception, host, statement)) {
+                return;
+            }
+            RollingCount hostCount = getOrCreateCount(host);
+            hostCount.increment();
+        }
+
+        boolean isExcluded(Host host) {
+            Long excludedTime = exclusionTimes.get(host);
+            boolean expired = excludedTime != null && clock.nanoTime() - excludedTime >= retryPeriodNanos;
+            if (excludedTime == null || expired) {
+                if (maybeExcludeNow(host, excludedTime)) {
+                    return true;
+                }
+                if (expired) {
+                    // Cleanup, but make sure we don't overwrite if another thread just set it
+                    exclusionTimes.remove(host, excludedTime);
+                }
+                return false;
+            } else { // host is already excluded
+                return true;
+            }
+        }
+
+        // Exclude if we're over the threshold
+        private boolean maybeExcludeNow(Host host, Long previousTime) {
+            RollingCount rollingCount = getOrCreateCount(host);
+            long count = rollingCount.get();
+            if (count > maxErrorsPerMinute) {
+                excludeNow(host, count, previousTime);
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        // Set the exclusion time to now, handling potential races
+        private void excludeNow(Host host, long count, Long previousTime) {
+            long now = clock.nanoTime();
+            boolean didNotRace = (previousTime == null)
+                    ? exclusionTimes.putIfAbsent(host, now) == null
+                    : exclusionTimes.replace(host, previousTime, now);
+
+            if (didNotRace && logger.isDebugEnabled()) {
+                logger.debug(String.format("Host %s encountered %d errors in the last minute, which is more " +
+                                "than the maximum allowed (%d). It will be excluded from query plans for the " +
+                                "next %d nanoseconds.",
```

Comment:
It might be more useful to convert the logged value into milliseconds so it is easier for users to understand.